frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -5156,6 +5399,343 @@ const TypedFnSignature* tryResolveDeinit(Context* context,
+  std::vector<ApplicabilityResult> rejected;
+  std::vector<const TypedFnSignature*> ambiguous;
+
+  std::vector<CallInfoActual> actuals;
+  for (int i = 0; i < tfs->numFormals(); i++) {
+    auto decl = tfs->untyped()->formalDecl(i);
+    auto name = UniqueString();
+    if (auto formal = decl->toFormal()) {
+        name = formal->name();
+    } else if (decl->isVarArgFormal()) {
+      CHPL_UNIMPL("vararg formals in interface function requirements");
+      return ID();
+    }
+    actuals.emplace_back(tfs->formalType(i), name);
+  }
+
+  CallInfo ci {
+      tfs->untyped()->name(),
+      /* calledType */ QualifiedType(),
+      /* isMethodCall */ fn->isMethod(),
+      /* hasQuestionArg */ false,
+      /* isParenless */ fn->isParenless(),
+      std::move(actuals)
+  };
+
+  // TODO: how to note this?
+  auto c =
+    resolveGeneratedCall(rc->context(), fn, ci, inScopes);
+
+  bool failed = c.exprType().isUnknownOrErroneous();
+  if (failed && fn->body()) {
+    // template has a default implementation; return it, we're good.
+    return fn->id();
+  } else if (failed && c.mostSpecific().isAmbiguous()) {
+    // TODO: no way at this time to collected candidates rejected due to
+    //       ambiguity, so just report an empty list.
+    CHPL_REPORT(rc->context(), InterfaceAmbiguousFn, iftForErr, implPointIdForErr,
+                fn, std::move(ambiguous));
+    return ID();
+  } else if (failed) {
+    // Failed to find a call, not due to ambiguity. Re-run call and gather
+    // rejected candidates.
+    resolveGeneratedCall(rc->context(), fn, ci, inScopes, &rejected);
+    CHPL_REPORT(rc->context(), InterfaceMissingFn, iftForErr, implPointIdForErr,
+                tfs, ci, std::move(rejected));
+    return ID();
+  }
+
+  CHPL_ASSERT(!failed);
+
+  const TypedFnSignature* foundFn = nullptr;
+  if (c.mostSpecific().numBest() > 1) {
+    CHPL_UNIMPL("return intent overloading in interface constraint checking");
+    return ID();
+  } else {
+    // There's only one function; we should still check its intent.
+    foundFn = c.mostSpecific().only().fn();
+    auto foundIntent = parsing::idToFnReturnIntent(rc->context(), foundFn->id());
+
+    if (foundIntent == fn->returnIntent()) {
+      // fine
+    } else {
+      CHPL_REPORT(rc->context(), InterfaceInvalidIntent, iftForErr, implPointIdForErr,
+                  tfs, foundFn);
+      return ID();
+    }
+  }
+
+  // Validate that the formal names are in the right order. We could do this
+  // by resolving a generated call with every actual being named, but this
+  // will miss some cases, like when the template is:
+  //
+  //      proc f(x: int, y: int)
+  //
+  //  and the actual function is:
+  //
+  //      proc f(y: int, x: int):
+  //
+  // such calls would resolve match with and without named actuals, but the
+  // actuals are not in the right order compared to the template.
+  FormalActualMap faMap(foundFn, ci);
```

Comment:
Function types would probably like this as well, so we could make this a helper.

Reply:
I shall leave extracting the helper for a follow-up